### PR TITLE
Option to force non xc8-cc

### PR DESCRIPTION
**Pull request by aytey**
I am currently working with some xc8 code that _does not_ compile with `xc8-cc`, but does with `xc8`. This PR adds a CMake ~variable~ option (`MICROCHIP_XC8_FORCE_NON_CC`) that allows the user to force the non-`cc` compiler.

To use it, add

```
set(MICROCHIP_XC8_FORCE_NON_CC ON CACHE BOOL "" FORCE)
```

to the top of your CMakeLists.txt, or provide it on the command line:

```
-DMICROCHIP_XC8_FORCE_NON_CC:BOOL=ON
```

### CODE REVIEW

File: Modules/Platform/MicrochipMCU-C-XC8.cmake [hunk 1]
```diff
@@ -30,11 +30,15 @@ endif()
 
 set(CMAKE_FIND_ROOT_PATH "${MICROCHIP_XC8_PATH}")
 
-# skip compiler search and just use XC8
-find_program(CMAKE_C_COMPILER "xc8-cc"
-    PATHS "${MICROCHIP_XC8_PATH}"
-    PATH_SUFFIXES "bin"
-)
+if(NOT MICROCHIP_XC8_FORCE_NON_CC)
+    if(NOT CMAKE_C_COMPILER)
```
Comment (enbyted) [inline Modules/Platform/MicrochipMCU-C-XC8.cmake]: `if(NOT CMAKE_C_COMPILER)` should not be required here, why did you add it?

**Comment by enbyted** [comment]
Please add appropriate `set([...] CACHE [...])` so that users can find this variable in cache/using GUI.
Besides that It's good change, thanks for that!

File: Modules/Platform/MicrochipMCU-C-XC8.cmake (hunk 1 above)
Comment (aytey) [reply Modules/Platform/MicrochipMCU-C-XC8.cmake]: Hmm, well, I think this allows people to set `MICROCHIP_XC8_FORCE_NON_CC` in their CMakeLists.txt *and* override it from the outside, right?

The logic is the same as a few lines further down (already there), where it checks ` if(NOT CMAKE_C_COMPILER)`.

Happy to remove this if we feel that setting `MICROCHIP_XC8_FORCE_NON_CC` should nuke `CC`/`CXX`.

**Comment by aytey** [comment]
> Please add appropriate set([...] CACHE [...]) so that users can find this variable in cache/using GUI.

Oops! Will do!

File: Modules/Platform/MicrochipMCU-C-XC8.cmake (hunk 1 above)
Comment (enbyted) [reply Modules/Platform/MicrochipMCU-C-XC8.cmake]: > Hmm, well, I think this allows people to set MICROCHIP_XC8_FORCE_NON_CC in their CMakeLists.txt and override it from the outside, right?

No, this is called from toolchain file, so you can set this variable in 3 places:
 - `CMakeCache.txt`
 - Command line
 - `PreLoad.cmake`

*EDIT* I quess, you can also set it before calling `project()` in your CMakeLists, that should also work.

> The logic is the same as a few lines further down (already there), where it checks `if(NOT CMAKE_C_COMPILER)`.

Yes, but here this is first `find_program` in the file, thus `CMAKE_C_COMPILER` should be not set anyway. I'm pretty sure that if it was set before CMake would not execute toolchain file in the first place.

It shouldn't cause any harm, but it wasn't in the original code and I'm 99% positive that it's not needed (though I'm hardly CMake expert...).

> Happy to remove this if we feel that setting MICROCHIP_XC8_FORCE_NON_CC should nuke `CC`/`CXX`.

I don't think that it will nuke anything because there's nothing to nuke at this point - toolchain file is supposed to find the toolchain and tell CMake how to use it. This variable should just change this module's behavior. 

If it still works for you - please remove it.

Comment (aytey) [reply Modules/Platform/MicrochipMCU-C-XC8.cmake]: Should be done!

**Comment by aytey** [comment]
> Please add appropriate set([...] CACHE [...]) so that users can find this variable in cache/using GUI.

I took a different approach and made this an option.

- [x] Do we want to document how to set this? Right now, you need to do this:

```
set(MICROCHIP_XC8_FORCE_NON_CC ON CACHE BOOL "" FORCE)
```

or pass `-DMICROCHIP_XC8_FORCE_NON_CC:BOOL=ON`.

**Comment by aytey** [comment]
Screenshot 2020-11-03 at 10 45 22

**Comment by enbyted** [comment]
> Do we want to document how to set this? Right now, you need to do this:
> ```
> set(MICROCHIP_XC8_FORCE_NON_CC ON CACHE BOOL "" FORCE)
> ```

IMHO it's not the correct way to do that (and you need to set it before `project()` call, I think). You should pass `-DMICROCHIP_XC8_FORCE_NON_CC:BOOL=ON` (or use gui to do it for you). But if you feel that it would help others than feel free to update the readme.

**Comment by aytey** [comment]
Interestingly, in my CMakeLists, I *do* have it before `project`, but if I do:

```
set(MICROCHIP_XC8_FORCE_NON_CC ON)
```

then I get:

```
CMake Warning (dev) at external/cmake-microchip/Modules/Platform/MicrochipMCU-C-XC8.cmake:22 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'MICROCHIP_XC8_FORCE_NON_CC'.
Call Stack (most recent call first):
  external/cmake-microchip/Modules/Platform/MicrochipMCU-C.cmake:18 (include)
  C:/cmake/cmake-3.18.4-win64-x64/share/cmake-3.18/Modules/CMakeDetermineCCompiler.cmake:28 (include)
  CMakeLists.txt:17 (project)
```

but with:

```
set(MICROCHIP_XC8_FORCE_NON_CC ON CACHE BOOL "" FORCE)
```

everything works.

Maybe I should just document the `-D` (on the command line) but leave doing in CMakeLists as an "exercise for the reader"?

**Comment by enbyted** [comment]
Oh, right, that's why all others are normal variables.

> Maybe I should just document the -D (on the command line) but leave doing in CMakeLists as an "exercise for the reader"?

yes, that should be the preferred method anyway.

**Comment by aytey** [comment]
> yes, that should be the preferred method anyway.

Done!

**Comment by enbyted** [comment]
Merged, thanks for your work!

**Comment by aytey** [comment]
No worries 👍

**Comment by alxrMironov** [comment]
The "issue" feature is disabled by default in forked repo.
Also for you know - i am pretty new with build systems and may have mistakes.

The "Non CC" feature seems to be broken, there is no way to invoke xc8-cc.exe after the last pull request in @enbyted repo.
The xc8.exe alway invoked, with or without flag.
When im replace toolchain to one from root Elemecca repo, it works as expected (xc8-cc.exe invoked).

[https://ibb.co/8cnFJwS](url)

**Comment by aytey** [comment]
@alxrMironov did you make and then delete a comment? Did you resolve your issue?

**Comment by aytey** [comment]
Either way, I cannot reproduce your issue:

```
avj@xc8vm MSYS ~/xc8_build/build
# rm -rf * && cmake -G "MSYS Makefiles" ..
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8-cc.exe
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8-cc.exe - works
-- Configuring done
-- Generating done
-- Build files have been written to: C:/msys64/home/avj/xc8_build/build

avj@xc8vm MSYS ~/xc8_build/build
# rm -rf * && cmake -G "MSYS Makefiles" -DMICROCHIP_XC8_FORCE_NON_CC:BOOL=ON ..
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8.exe
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8.exe - works
-- Configuring done
-- Generating done
-- Build files have been written to: C:/msys64/home/avj/xc8_build/build

avj@xc8vm MSYS ~/xc8_build/build
# rm -rf * && cmake -G "MSYS Makefiles" -DMICROCHIP_XC8_FORCE_NON_CC:BOOL=OFF ..
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8-cc.exe
-- Check for working C compiler: C:/Microchip/xc8/v2.30/bin/xc8-cc.exe - works
-- Configuring done
-- Generating done
-- Build files have been written to: C:/msys64/home/avj/xc8_build/build

```

**Comment by enbyted** [comment]
Orignal comment by @alxrMironov :
> The "issue" feature is disabled by default in forked repo.
> Also for you know - i am pretty new with build systems and may have mistakes.
> 
> The "Non CC" feature seems to be broken, there is no way to invoke xc8-cc.exe after the last pull request in @enbyted repo.
> The xc8.exe alway invoked, with or without flag.
> When im replace toolchain to one from root Elemecca repo, it works as expected (xc8-cc.exe invoked).
> 
> https://ibb.co/8cnFJwS

> The "issue" feature is disabled by default in forked repo.

I didn't realize that, it's enabled now

> The xc8.exe alway invoked, with or without flag.

I'll test in on Windows in powershell on Monday, but it should work as @andrewvaughanj provided above.

> When im replace toolchain to one from root Elemecca repo, it works as expected (xc8-cc.exe invoked).

Original repository doesn't even recognize xc8-cc, I don't think that's possible.

Could you please open an issue now with more details about your setup, so we can look into it in more detail?

**Comment by aytey** [comment]
It looks good to me ...

Screenshot 2020-11-28 at 15 19 54
